File: src/main/java/com/datarank/tempest/allocator/model/ModelCluster.java
/* The MIT License (MIT)
 * Copyright (c) 2015 DataRank, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package com.datarank.tempest.allocator.model;

import com.datarank.tempest.allocator.RandomList;
import org.elasticsearch.cluster.ClusterInfo;
import org.elasticsearch.cluster.InternalClusterInfoService;
import org.elasticsearch.cluster.routing.MutableShardRouting;
import org.elasticsearch.cluster.routing.RoutingNode;
import org.elasticsearch.cluster.routing.allocation.RoutingAllocation;
import org.elasticsearch.common.base.Predicate;
import org.elasticsearch.common.settings.Settings;
import org.elasticsearch.index.shard.ShardId;

import java.util.*;

/**
 * Represents a possible state of the cluster. Forking is done by new instance creation, rather than in place. In place
 * might be faster, but it would be a lot more likely to run into validity issues, particularly with AllocationDeciders.
 */
public class ModelCluster {
    public static Random RANDOM = new Random();

    private static final Predicate<MutableShardRouting> assignedPredicate = new Predicate<MutableShardRouting>() {
        @Override
        public boolean apply(MutableShardRouting input) {
            return input.assignedToNode();
        }
    };
    private final RoutingAllocation routingAllocation;
    private final ClusterInfo clusterInfo;
    private final Settings settings;
    private final ModelCluster startingCluster;

    private ModelNodes modelNodes;
    // partitioning shards to avoid the expense of filtering before random selection from the collection of shards
    private List<ModelShard> unassignedShards;
    private List<ModelAllocationDecider> allocationDeciders;
    private List<ModelOperation> forkingOperationHistory;

    //region Constructors
    //TODO:SCA: remove testing constructor and mock out RoutingAllocation
    public ModelCluster(final ClusterInfo clusterInfo, final ModelNodes modelNodes,
                        final List<ModelShard> unassignedShards, final List<ModelAllocationDecider> allocationDeciders,
                        List<ModelOperation> forkingOperationHistory, Settings settings, Random random) {
        this.clusterInfo = clusterInfo;
        this.modelNodes = modelNodes;
        this.unassignedShards = unassignedShards;
        this.allocationDeciders = allocationDeciders;
        this.forkingOperationHistory = forkingOperationHistory;
        this.startingCluster = this;
        this.settings = settings;
        RANDOM = random;

        routingAllocation = null;
    }

    /**
     * Copy constructor, shallow copies all collection elements except for modelNode_s.
     * @param other
     */
    public ModelCluster(final ModelCluster other) {
        this.routingAllocation = other.getRoutingAllocation();
        this.clusterInfo = other.getClusterInfo();
        // modelNodes need to be deep copied to prevent accidental modification of a cluster by its copy
        this.modelNodes = new ModelNodes(other.getModelNodes(), RANDOM);
        // each time we fork, move all ignoredUnassigned shards back to unassigned
        this.unassignedShards = new RandomList<>(other.getUnassignedShards(), RANDOM);
        this.forkingOperationHistory = new ArrayList<>(other.getForkingOperationHistory());
        this.settings = other.getSettings();
        this.allocationDeciders = other.getAllocationDeciders();
        this.startingCluster = other.getStartingCluster();
    }

    public ModelCluster(final RoutingAllocation allocation, final Settings settings) {
        this.routingAllocation = allocation;
        this.clusterInfo = allocation.clusterInfo();
        this.forkingOperationHistory = new ArrayList<>();
        this.settings = settings;
        this.startingCluster = this;

        buildModel();
    }

    public ModelCluster(final RoutingAllocation allocation, final Settings settings, final Random random) {
        this.routingAllocation = allocation;
        this.clusterInfo = allocation.clusterInfo();
        this.forkingOperationHistory = new ArrayList<>();
        this.settings = settings;
        this.startingCluster = this;
        RANDOM = random;

        buildModel();
    }
    //endregion

    //region Forking

    /**
     * Attempts to allocate all unassigned shards in the cluster.
     *
     * Note that because this method randomly chooses an
     * unassigned shard to allocate and the node to allocate it on, and because the ability of a shard to be allocated
     * on a node is effected by previous allocations, it is possible that there exists one or more valid end clusters
     * where all shards are allocated, but that this method will not find one of them. The probability of finding a
     * valid end cluster state increases linearly with the number of valid end cluster states. The following will all
     * increase the probability of finding a valid end cluster state:
     * - Fewer routing filters/awareness restrictions. If none, this method is VERY likely to find a valid end cluster state.
     * - More nodes
     * - Fewer replicas
     *
     * @return A cluster with as many shards allocated as possible
     */
    public ModelCluster allocateUnassigned() {
        if (!this.hasUnassignedShards()) {
            return null;
        }
        ModelCluster forkedCluster = this;
        ModelCluster nextCluster;
        while (forkedCluster.hasUnassignedShards()) {
            nextCluster = forkedCluster.forkUnassigned();
            if (nextCluster == forkedCluster) {
                return forkedCluster;
            }
            forkedCluster = nextCluster;
        }
        return forkedCluster;
    }

    public ModelCluster rebalance() {
        if (this.hasUnassignedShards()) {
            throw new IllegalStateException("Cannot rebalance while there are still unassigned shards.");
        }
        if (!modelNodes.hasShards()) {
            throw new IllegalStateException("Cannot relocate a random shard if there are no shards to relocate.");
        }
        return forkAssigned();
    }

    public ModelCluster move(final ModelShard shard, final ModelNode sourceNode) {
        if (!modelNodes.isAssigned(shard) || !modelNodes.nodeContains(sourceNode.getNodeId(), shard)) {
            return this;
        }
        ModelOperation operation = findValidRelocation(shard, sourceNode);
        // no possible moves for this shard, leave it where it is
        if (operation == ModelOperation.NO_OP) {
            return this;
        }
        return forkCluster(this, operation);
    }

    /**
     * Try to fork the cluster by moving unassigned shards to a node
     * @return a new ModelCluster with the move performed, or null if no such move is possible
     */
    private ModelCluster forkUnassigned() {
        if (unassignedShards.isEmpty()) { return null; }
        ModelOperation operation;
        for (ModelShard unassignedShard : unassignedShards) {
            operation = findValidRelocation(unassignedShard);
            if (operation == ModelOperation.NO_OP) { continue; }
            return forkCluster(this, operation);
        }
        return this;
    }

    /**
     * Try to fork the cluster by moving a shard that is already assigned to a node
     * @return a new ModelCluster with the move performed, or null if no such move is possible
     */
    private ModelCluster forkAssigned() {
        ModelOperation operation;
        for (ModelNode sourceNode : modelNodes) {
            if (sourceNode.getShards().isEmpty()) { continue; }
            for (ModelShard shard : sourceNode.getShards()) {
                operation = tryMoveToNode(shard, sourceNode);
                // no possible moves for this shard, leave it where it is
                if (operation == ModelOperation.NO_OP) { continue; }
                return forkCluster(this, operation);
            }
        }
        return this;
    }

    /**
     * Forks sourceCluster by copying it and executing operation on the new cluster. To avoid redundancy,
     * no validation checks are performed on operation.
     * @param sourceCluster
     * @param operation
     * @return a new ModelCluster representing sourceCluster modified by operation.
     */
    private static ModelCluster forkCluster(final ModelCluster sourceCluster, final ModelOperation operation) {
        ModelCluster forkedCluster = new ModelCluster(sourceCluster);
        if (!forkedCluster.executeOperation(operation)) {
            throw new RuntimeException("Failed to execute operation on the cluster.");
        }
        return forkedCluster;
    }

    /**
     * Attempts to execute an operation on the current ModelCluster.
     * @param operation
     * @return true if the operation was successful, false otherwise.
     */
    public boolean executeOperation(final ModelOperation operation){
        if (operation == ModelOperation.NO_OP) {
            return true;
        }
        switch (operation.sourceType) {
            case UNASSIGNED:
                if (!unassignedShards.remove(operation.modelShard)) { return false; }
                if (!modelNodes.addShard(operation.destinationNode.getNodeId(), operation.modelShard)) { return false; }
                break;
            case NODE:
                if (!modelNodes.removeShard(operation.sourceNode.getNodeId(), operation.modelShard)) { return false; }
                if (!modelNodes.addShard(operation.destinationNode.getNodeId(), operation.modelShard)) { return false; }
                break;
            default:
                throw new IllegalStateException("Invalid sourceType " + operation.sourceType + " for operation.");
        }
        return this.forkingOperationHistory.add(operation);
    }

    /**
     * Determines if an operation is valid given the current ModelCluster. Provides some verification that the state is
     * valid and stable.
     * @param operation
     * @return
     */
    private boolean canExecute(final ModelOperation operation) {
        if (operation == ModelOperation.NO_OP) {
            return true;
        }

        ModelShard shard = operation.modelShard;
        if (modelNodes.nodeContains(operation.destinationNode.getNodeId(), shard)) { return false; }
        switch (operation.sourceType) {
            case UNASSIGNED:
                if (!unassignedShards.contains(shard)) { return false; }
                if (modelNodes.isAssigned(shard)) { return false; }
                break;
            case NODE:
                if (unassignedShards.contains(shard)) { return false; }
                if (!modelNodes.nodeContains(operation.sourceNode.getNodeId(), shard)) { return false; }
                if (modelNodes.nodeContains(operation.destinationNode.getNodeId(), shard)) { return false; }
                break;
            default:
                throw new IllegalStateException("Invalid sourceType " + operation.sourceType + " for operation.");
        }
        for (ModelAllocationDecider allocationDecider : allocationDeciders) {
           if (!allocationDecider.canRelocate(this, operation)) { return false; }
        }
        return true;
    }

    /**
     * Finds the smallest node that the unassigned shard can be moved to.
     * @param shard
     * @return a ModelOperation representing a move from UNASSIGNED to the destination node.
     */
    private ModelOperation findValidRelocation(final ModelShard shard) {
        ModelOperation operation;
        for (ModelNode destinationNode : modelNodes.getNodesSortedBySize()) {
            operation = new ModelOperation(shard, null, destinationNode);
            if (ModelOperation.isValid(operation) && canExecute(operation)) {
                return operation;
            }
        }
        return ModelOperation.NO_OP;
    }

    /**
     * Randomly finds a possible destination for a shard.
     *
     * Unassigned shards can be moved to a node or, if not possible, to ignoredUnassigned.
     * Shards on nodes (assigned) can be moved to another node. If not possible, ModelOperation.NO_OP is returned,
     * representing that the shard must stay in its current location.
     * @param shard
     * @return a ModelOperation representing a move from the shard's current location to a destination node
     */
    private ModelOperation findValidRelocation(final ModelShard shard, final ModelNode sourceNode) {
        ModelOperation operation = tryMoveToNode(shard, sourceNode);
        return operation;
    }

    private ModelOperation tryMoveToNode(final ModelShard shard, final ModelNode sourceNode) {
        ModelOperation operation;
        for (ModelNode destinationNode : modelNodes) {
            operation = new ModelOperation(shard, sourceNode, destinationNode);
            if (ModelOperation.isValid(operation) && canExecute(operation)) {
                return operation;
            }
        }
        return ModelOperation.NO_OP;
    }
    //endregion

    //region ElasticSearch Integration
    /**
     * Builds a complete model (cluster, nodes, and shards) from this.routingAllocation. Should only be called before
     * operating on the cluster.
     */
    private void buildModel() {
        this.modelNodes = new ModelNodes(RANDOM);
        this.unassignedShards = new RandomList<>(routingAllocation.routingNodes().unassigned().size(), RANDOM);
        this.allocationDeciders = buildDefaultAllocationDeciders(this.settings);
        // build ModelNodes
        for (RoutingNode routingNode : routingAllocation.routingNodes()) {
            modelNodes.add(new ModelNode(routingNode, RANDOM));
        }
        /* There's no simple way to get only primary/replica shards, and there's also no simple way to get the primary
           that corresponds to a replica. So build and add shards where they should go, then set primaries.
         */
        Map<ShardId, ModelShard> primaryShards = new HashMap<>();
        Map<ShardId, ModelShard> replicaShards = new HashMap<>();
        for (MutableShardRouting routingShard : routingAllocation.routingNodes().shards(assignedPredicate)){
            ModelShard shard = new ModelShard(routingShard, getShardSize(routingShard));
            if (routingShard.primary()){
                primaryShards.put(shard.getShardIdentifier(), shard);
                shard.setPrimaryShard(shard);
            }
            else {
                replicaShards.put(shard.getShardIdentifier(), shard);
            }
            modelNodes.addShard(routingShard.currentNodeId(), shard);
        }
        for (MutableShardRouting routingShard : routingAllocation.routingNodes().unassigned()) {
            ModelShard shard = new ModelShard(routingShard, getShardSize(routingShard));
            if (routingShard.primary()){
                primaryShards.put(shard.getShardIdentifier(), shard);
                shard.setPrimaryShard(shard);
            }
            else {
                replicaShards.put(shard.getShardIdentifier(), shard);
            }
            unassignedShards.add(shard);
        }

        for (MutableShardRouting routingShard : routingAllocation.routingNodes().ignoredUnassigned()) {
            ModelShard shard = new ModelShard(routingShard, getShardSize(routingShard));
            if (routingShard.primary()) {
                primaryShards.put(shard.getShardIdentifier(), shard);
                shard.setPrimaryShard(shard);
            }
            else {
                replicaShards.put(shard.getShardIdentifier(), shard);
            }
            unassignedShards.add(shard);
        }

        // set primaries
        for (ModelShard replica : replicaShards.values()) {
            replica.setPrimaryShard(primaryShards.get(replica.getShardIdentifier())); // replicas and primaries share ids
        }
    }

    private List<ModelAllocationDecider> buildDefaultAllocationDeciders(final Settings settings) {
        List<ModelAllocationDecider> deciders = new ArrayList<>();
        deciders.add(new ModelSameShardAllocationDecider(settings));
        deciders.add(new ModelEnableAllocationDecider(settings));
        deciders.add(new ModelRebalanceOnlyWhenActiveAllocationDecider(settings));
        deciders.add(new ModelReplicaAfterPrimaryActiveAllocationDecider(settings));
        return deciders;
    }
    //endregion

    //region Calculations and Sizes
    public int getNumNodes() {
        return modelNodes.size();
    }

    public int getNumShards() { return getNumAssignedShards() + getNumUnassignedShards(); }

    public int getNumAssignedShards() {
        return modelNodes.getNumShards();
    }

    public int getNumUnassignedShards() { return unassignedShards.size(); }

    public long getTotalShardsSize() {
        return modelNodes.getShardsSize() + calculateTotalSizeOfShards(unassignedShards);
    }

    private long calculateTotalSizeOfShards(final Collection<ModelShard> shards) {
        long size = 0l;
        for (ModelShard shard : shards) {
            size += shard.getSize();
        }
        return size;
    }

    private long getShardSize(final MutableShardRouting routingShard) {
        String key = InternalClusterInfoService.shardIdentifierFromRouting(routingShard);

        if (!clusterInfo.getShardSizes().containsKey(key)) {
            return ModelShard.UNKNOWN_SHARD_SIZE;
        }

        return clusterInfo.getShardSizes().get(key);
    }

    public static double stddev(final ModelCluster cluster) {
        double mean = cluster.getTotalShardsSize() / cluster.getNumNodes();
        double variance = 0d;

        for (ModelNode node : cluster.getModelNodes()) {
            double var = node.getTotalShardSize() - mean;
            variance += var * var;
        }
        return Math.sqrt(variance / cluster.getNumNodes());
    }

    /**
     * Calculates the difference between the largest node and the smallest node
     * @param cluster
     * @return
     */
    public static long sizeRange(final ModelCluster cluster) {
        long max = Long.MIN_VALUE;
        long min = Long.MAX_VALUE;
        long size;
        for (ModelNode node : cluster.getModelNodes()) {
            size = node.getTotalShardSize();
            if (size > max) { max = size; }
            if (size < min) { min = size; }
        }
        return max - min;
    }
    //endregion

    //region Accessors
    public RoutingAllocation getRoutingAllocation() {
        return routingAllocation;
    }

    public ClusterInfo getClusterInfo() {
        return clusterInfo;
    }

    public ModelNodes getModelNodes() {
        return modelNodes;
    }

    public List<ModelShard> getUnassignedShards() { return unassignedShards; }

    public boolean hasUnassignedShards() { return !unassignedShards.isEmpty(); }

    public List<ModelAllocationDecider> getAllocationDeciders() { return allocationDeciders; }

    public List<ModelOperation> getForkingOperationHistory() { return forkingOperationHistory; }

    public ModelCluster getStartingCluster() {
        return startingCluster;
    }

    public Settings getSettings() {
        return settings;
    }
    //endregion

    public void clearForkingOperationHistory() {
        forkingOperationHistory.clear();
    }
}

File: src/main/java/com/simplymeasured/elasticsearch/plugins/tempest/balancer/ModelCluster.kt
/*
 * The MIT License (MIT)
 * Copyright (c) 2016 DataRank, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 *  of this software and associated documentation files (the "Software"), to deal
 *  in the Software without restriction, including without limitation the rights
 *  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 *  copies of the Software, and to permit persons to whom the Software is
 *  furnished to do so, subject to the following conditions:
 *
 *  The above copyright notice and this permission notice shall be included in
 *  all copies or substantial portions of the Software.
 *
 *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 *  THE SOFTWARE.
 *
 */

package com.simplymeasured.elasticsearch.plugins.tempest.balancer

import org.eclipse.collections.api.list.ListIterable
import org.eclipse.collections.api.list.MutableList
import org.eclipse.collections.api.set.SetIterable
import org.eclipse.collections.api.tuple.Pair
import org.eclipse.collections.impl.block.factory.Comparators
import org.eclipse.collections.impl.block.factory.Functions
import org.eclipse.collections.impl.factory.Sets
import org.eclipse.collections.impl.list.mutable.FastList
import org.eclipse.collections.impl.tuple.Tuples
import org.eclipse.collections.impl.utility.LazyIterate
import org.elasticsearch.cluster.ClusterInfo
import org.elasticsearch.cluster.routing.RoutingNode
import org.elasticsearch.cluster.routing.RoutingNodes
import org.elasticsearch.cluster.routing.ShardRouting
import org.elasticsearch.cluster.routing.ShardRoutingState
import java.util.*

/**
 * Represents a possible cluster future easy/cheap function for manipulation
 */
class ModelCluster private constructor(
        val modelNodes: ListIterable<ModelNode>,
        val mockDeciders: ListIterable<MockDecider>,
        val shardSizeCalculator: ShardSizeCalculator,
        val expungeBlacklistedNodes: Boolean,
        val random: Random) {

    /**
     * main constructor used for creating a base model from ES provided data
     */
    constructor(routingNodes: RoutingNodes,
                blacklistFilter: (RoutingNode) -> Boolean,
                shardSizeCalculator: ShardSizeCalculator,
                mockDeciders: ListIterable<MockDecider>,
                expungeBlacklistedNodes: Boolean,
                random: Random) :
        this (modelNodes = LazyIterate.adapt(routingNodes)
                                      .collect { it -> ModelNode.buildModelNode(
                                              node = it,
                                              expungeBlacklistedNodes = expungeBlacklistedNodes,
                                              blacklistFilter = blacklistFilter,
                                              shardSizeCalculator = shardSizeCalculator) }
                                      .toList(),
              mockDeciders = mockDeciders,
              shardSizeCalculator = shardSizeCalculator,
              expungeBlacklistedNodes = expungeBlacklistedNodes,
              random = random)

    /**
     * deep copy constructor
     */
    constructor(other: ModelCluster) :
        this(modelNodes = other.modelNodes.collect(::ModelNode),
             mockDeciders = other.mockDeciders,
             shardSizeCalculator = other.shardSizeCalculator,
             expungeBlacklistedNodes = other.expungeBlacklistedNodes,
             random = other.random)

    companion object {
        val MAX_MOVE_ATTEMPTS = 1000
    }

    val sourceNodes: ListIterable<ModelNode> = if (expungeableShardsExist()) modelNodes.select { it.isExpunging && it.shards.notEmpty() }
                                               else                          modelNodes.select { !it.isBlacklisted && it.shards.notEmpty()}

    val destinationNodes: ListIterable<ModelNode> = modelNodes.select { !it.isBlacklisted }
    val expectedUnitCapacity: Double = calculateExpectedUnitCapacity()

    private fun expungeableShardsExist() = modelNodes.select { it.isExpunging }
                                                     .anySatisfy { it.shards.isNotEmpty() }

    /**
     * Attempt to make a random move that attempts to satisfy all mock deciders
     */
    fun makeRandomMove(previousMoves: Collection<MoveAction>): MoveAction? {
        for (attempt in 0..MAX_MOVE_ATTEMPTS) {
            val sourceNode = findValidSourceNode()
            if (sourceNode.shards.isEmpty()) { continue }

            val destNode = findValidDestinationNode()
            val shard = sourceNode.shards.get(random.nextInt(sourceNode.shards.size))

            if (mockDeciders.all { it.canMove(shard, destNode, previousMoves) }) {
                shard.state = ShardRoutingState.RELOCATING
                destNode.shards.add(shard.copy(state = ShardRoutingState.INITIALIZING))
                return MoveAction(sourceNode, shard, destNode)
            }
        }

        return null
    }

    private fun findValidDestinationNode() = destinationNodes.get(random.nextInt(destinationNodes.size()))

    private fun findValidSourceNode() = sourceNodes.get(random.nextInt(sourceNodes.size()))

    fun applyMoveChain(moveChain: MoveChain) {
        moveChain.moveBatches.forEach { applyMoveActionBatch(it) }
    }

    fun applyMoveActionBatch(moveActionBatch: MoveActionBatch) {
        moveActionBatch.moves.forEach { applyMoveAction(it) }
    }

    fun applyMoveAction(moveAction: MoveAction) {
        // we have to localize a move action since model nodes and shards are owned by their respective model clusters;
        // this method is not fast O(n) in terms of shard and node searching but is also used in non critical paths
        val localSourceNode = modelNodes.find { it.nodeId == moveAction.sourceNode.nodeId }!!
        val localDestNode = modelNodes.find { it.nodeId == moveAction.destNode.nodeId }!!
        val localShard = localSourceNode.shards.find { it.backingShard.isSameShard(moveAction.shard.backingShard) }!!

        localSourceNode.shards.remove(localShard)
        localDestNode.shards.add(localShard)
    }

    fun applyShardInitialization(shard: ShardRouting, allocatedNode: ModelNode) {
        val modelShard = ModelShard(shard, size = shardSizeCalculator.actualShardSize(shard), estimatedSize = shardSizeCalculator.estimateShardSize(shard))
        modelShard.state = ShardRoutingState.INITIALIZING
        allocatedNode.shards.add(modelShard)
    }

    /**
     * During a model move the cluster will have some shares in two places: the source in a state of relocating and the dest
     * in a state of initialization. This method will remove shares in the relocating state and "start" initializing shards
     */
    fun stabilizeCluster() {
        modelNodes.forEach { it.stabilizeNode() }
    }

    /**
     * Calculate the balance score for the entire cluster
     *
     * NOTE: for blacklisted nodes while an expunge is happening we treat their capacity as 0 since we are trying to
     *       move everything off of them
     */
    fun calculateBalanceScore(): Double = modelNodes.fold(
            0.0,
            { score, node -> score + node.calculateNodeScore(expectedUnitCapacity) })

    /**
     * Calculate the risk for the cluster.
     *
     * Note this is currently defined in terms of the size of the most over-capacity node in the cluster
     */
    fun calculateRisk(): Double = modelNodes
            .map { it.calculateNodeScore(expectedUnitCapacity) }
            .max() ?: 0.0

    private fun calculateExpectedUnitCapacity(): Double {
        val shards = modelNodes.flatMap { it.shards }
        val totalNodeSize = shards.map { it.estimatedSize }.filterNot { it < 0 }.sum().toDouble()
        val totalClusterCapacity = destinationNodes.map { it.allocationScale }.sum()
        return totalNodeSize / totalClusterCapacity
    }

    /**
     * Find the percent change of the node with the most usage change between this cluster and another cluster
     */
    fun findBestNodeUsageImprovement(otherCluster: ModelCluster): Double = LazyIterate
            .zip(modelNodes, otherCluster.modelNodes)
            .collect { Tuples.pair(it.one.calculateNormalizedNodeUsage(), it.two.calculateNormalizedNodeUsage()) }
            .collect { Math.abs(1.0 - safeRateDivide(it.two, it.one)) }
            .max()

    private fun safeRateDivide(top: Double, bottom: Double) : Double {
        if (top == 0.0 && bottom == 0.0) { return 1.0 }
        if (bottom == 0.0) { return Double.MAX_VALUE; }
        return top/bottom
    }

    fun findBestNodesForShard(unassignedShard: ShardRouting) : ListIterable<ModelNode> {
        val modelShard = ModelShard(unassignedShard, size = shardSizeCalculator.actualShardSize(unassignedShard), estimatedSize = shardSizeCalculator.estimateShardSize(unassignedShard))
        return destinationNodes.select { node -> mockDeciders.all { decider -> decider.canAllocate(modelShard, node) } }
                               .toSortedList(Comparators.chain(
                                       Comparators.byFunction<ModelNode, Double> {it.calculateNormalizedNodeUsage()},
                                       Comparators.byFunction<ModelNode, Int> {it.shards.size}))
    }

    override fun toString(): String {
        return modelNodes
                .flatMap { it.shards }
                .map { "${it.backingShard.currentNodeId()} ${it.backingShard.index} ${it.backingShard.id} ${it.state} (${it.estimatedSize}/${it.size})" }
                .toList()
                .sorted()
                .joinToString("\n")
    }
}

/**
 * Model representation of a Node and its shards
 */
class ModelNode private constructor(
        val backingNode: RoutingNode,
        val nodeId: String,
        val shards: MutableList<ModelShard>,
        val allocationScale: Double,
        val isBlacklisted: Boolean,
        val isExpunging: Boolean) {

    /**
     * Deep Copy Constructor
     */
    constructor(other: ModelNode) :
        this(other.backingNode,
             other.nodeId,
             other.shards.collect { it.copy() },
             other.allocationScale,
             other.isBlacklisted,
             other.isExpunging)

    /**
     * main constructor used when creating a base model from ES data structures
     */
    constructor(routingNode: RoutingNode,
                shardSizeCalculator: ShardSizeCalculator,
                isBlacklisted: Boolean,
                isExpunging: Boolean) :
        this(routingNode,
             routingNode.nodeId(),
             routingNode.copyShards().map { ModelShard(it, shardSizeCalculator.actualShardSize(it), shardSizeCalculator.estimateShardSize(it)) }.toFastList(),
             routingNode.node().attributes.getOrElse("allocation.scale", { "1.0" }).toDouble(),
             isBlacklisted,
             isExpunging)

    companion object {
        internal fun buildModelNode(
                node: RoutingNode,
                shardSizeCalculator: ShardSizeCalculator,
                blacklistFilter: (RoutingNode) -> Boolean,
                expungeBlacklistedNodes: Boolean): ModelNode {

            val blacklisted = blacklistFilter(node)
            return ModelNode(node, shardSizeCalculator, blacklisted, expungeBlacklistedNodes && blacklisted)
        }
    }

    /**
     * Calculate the node's wighted score (used for cluster risk and scoring)
     *
     * Note: currently using R^2 deviation from the norm as a score
     */
    fun calculateNodeScore(expectedUnitCapacity: Double): Double {
        val effectiveCapacity = if (isExpunging) 0.0 else expectedUnitCapacity
        return (effectiveCapacity - calculateNormalizedNodeUsage()).let { it * it }
    }

    /**
     * sum of all shard sizes
     */
    fun calculateUsage() : Long = shards.map { it.estimatedSize }.sum()

    /**
     * Like calculate usage but weighted in terms of the allocation scale used for this node
     */
    fun calculateNormalizedNodeUsage() : Double = calculateUsage()/allocationScale

    /**
     * Remove any relocating shards and set initializing shards to be started
     */
    fun stabilizeNode() {
        val shardIterator = shards.iterator()
        while (shardIterator.hasNext()) {
            val shard = shardIterator.next()
            when (shard.state) {
                ShardRoutingState.RELOCATING -> shardIterator.remove()
                ShardRoutingState.INITIALIZING -> shard.state = ShardRoutingState.STARTED
                else -> {}
            }
        }
    }
}

private fun <T> Iterable<T>.toFastList(): MutableList<T> = FastList.newList(this)

/**
 * Model representation of a shard
 */
data class ModelShard(val index: String, val id: Int, var state: ShardRoutingState, val primary: Boolean, val size: Long, val estimatedSize: Long, val backingShard: ShardRouting) {
    constructor(shardRouting: ShardRouting, size: Long, estimatedSize: Long) : this(shardRouting.index(), shardRouting.id(), shardRouting.state(), shardRouting.primary(), size, estimatedSize, shardRouting)
}
